### PR TITLE
feat: `Phone` add new package

### PR DESCRIPTION
**Pull request by vladimirpotekhin**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Part of #323

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/phone/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# @maskito/phone
+
+[![npm version](https://img.shields.io/npm/v/@maskito/phone.svg)](https://npmjs.com/package/@maskito/kit)
+[![npm bundle size](https://img.shields.io/bundlephobia/minzip/@maskito/phone)](https://bundlephobia.com/result?p=@maskito/kit)
```
Comment (nsbarsukov) [inline projects/phone/README.md:4]:
```suggestion
[![npm bundle size](https://img.shields.io/bundlephobia/minzip/@maskito/phone)](https://bundlephobia.com/result?p=@maskito/phone)
```

File: projects/phone/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# @maskito/phone
+
+[![npm version](https://img.shields.io/npm/v/@maskito/phone.svg)](https://npmjs.com/package/@maskito/kit)
```
Comment (nsbarsukov) [inline projects/phone/README.md:3]:
```suggestion
[![npm version](https://img.shields.io/npm/v/@maskito/phone.svg)](https://npmjs.com/package/@maskito/phone)
```

File: projects/phone/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+[![npm bundle size](https://img.shields.io/bundlephobia/minzip/@maskito/phone)](https://bundlephobia.com/result?p=@maskito/kit)
+
+<p align="center">
+    <img src="https://raw.githubusercontent.com/Tinkoff/maskito/main/projects/demo/src/assets/icons/maskito.svg" alt="Maskito logo" height="120px">
```
Comment (nsbarsukov) [inline projects/phone/README.md:7]:
```suggestion
    <img src="https://raw.githubusercontent.com/taiga-family/maskito/main/projects/demo/src/assets/icons/maskito.svg" alt="Maskito logo" height="120px">
```

File: projects/phone/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "description": "The optional framework-agnostic Maskito's package with phone masks",
+    "keywords": [
+        "mask",
+        "inputmask",
+        "phonemask",
+        "phone",
+        "javascript",
+        "typescript"
+    ],
+    "homepage": "https://tinkoff.github.io/maskito",
```
Comment (nsbarsukov) [inline projects/phone/package.json:13]:
```suggestion
    "homepage": "https://maskito.dev",
```

File: projects/phone/package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    "bugs": "https://github.com/tinkoff/maskito/issues",
```
Comment (nsbarsukov) [inline projects/phone/package.json:14]:
```suggestion
    "bugs": "https://github.com/taiga-family/maskito/issues",
```

File: projects/phone/package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    "repository": {
+        "type": "git",
+        "url": "https://github.com/tinkoff/maskito.git"
```
Comment (nsbarsukov) [inline projects/phone/package.json:17]:
```suggestion
        "url": "https://github.com/taiga-family/maskito.git"
```

File: projects/phone/package.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    "license": "Apache-2.0",
+    "author": {
+        "name": "Nikita Barsukov",
+        "email": "nikita.s.barsukov@gmail.com",
+        "url": "https://github.com/nsbarsukov"
```
Comment (nsbarsukov) [inline projects/phone/package.json:23]:
```suggestion
    "author": {
        "name": "Vladimir Potekhin",
        "email": "vladimir.potekh@gmail.com",
        "url": "https://github.com/vladimirpotekhin"
```

File: projects/phone/package.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        "Maksim Ivanov <splincodewd@yandex.ru>",
+        "Vladimir Potekhin <vladimir.potekh@gmail.com>",
+        "Nikita Barsukov <nikita.s.barsukov@gmail.com>"
+    ],
+    "devDependencies": {
+        "libphonenumber-js": "^1.10.37"
+    },
+    "peerDependencies": {
+        "@maskito/core": "^1.4.0",
+        "libphonenumber-js": "^1.10.37"
```
Comment (nsbarsukov) [inline projects/phone/package.json:36]: Should we set less strict constraint ?

Could it be `>=1.0.0` ?

File: projects/phone/src/lib/masks/phone/processors/cut-init-country-code-preprocessor.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            const newValue = `+${code} ${phone.nationalNumber}`;
+
+            return {
+                elementState: {
+                    value: newValue,
+                    selection,
+                },
+            };
+        } catch {
+            //
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/processors/cut-init-country-code-preprocessor.ts:44]:
```suggestion
            // TODO
```

File: projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+const MIN_LENGTH = 3;
+export function maskitoCutPhonePostprocessorGenerator(
+    metadata: MetadataJson,
+): MaskitoPostprocessor {
+    return ({value, selection}) => {
+        if (value.length > MIN_LENGTH) {
+            return {value: cutPhoneByValidLength(value, metadata), selection};
+        }
+
+        return {value, selection};
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts:15]: Should we use ternary operator ?

File: projects/demo/src/pages/pages.ts [hunk 11]
```diff
@@ -105,6 +105,12 @@ export const DEMO_PAGES: TuiDocPages = [
         route: DemoPath.DateTime,
         keywords: `date, day, month, year, mask, time, date-time, hour, minute, second, kit, generator`,
     },
+    {
+        section: 'Phone',
```
Comment (nsbarsukov) [inline projects/demo/src/pages/pages.ts:109]: What do you think about this idea ?

Screenshot 2023-08-11 at 17 28 58

Comment (nsbarsukov) [reply projects/demo/src/pages/pages.ts:109]: https://maskito.dev/addons/phone

File: projects/phone/src/lib/masks/phone/phone-mask.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
+import {MASKITO_DEFAULT_OPTIONS, MaskitoOptions} from '@maskito/core';
+import {maskitoCaretGuard, maskitoPrefixPostprocessorGenerator} from '@maskito/kit';
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/phone-mask.ts:2]: Should we add `@maskito/kit` to `peerDependencies`?

File: projects/phone/src/lib/masks/phone/utils/generate-phone-mask.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
+import {MaskExpression} from '@maskito/core/src/lib/types';
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/utils/generate-phone-mask.ts:1]:
```suggestion
import {MaskExpression} from '@maskito/core';
```

File: projects/core/package.json [hunk 14]
```diff
@@ -19,9 +19,9 @@
     },
     "license": "Apache-2.0",
     "author": {
-        "name": "Nikita Barsukov",
-        "email": "nikita.s.barsukov@gmail.com",
-        "url": "https://github.com/nsbarsukov"
+        "name": "Vladimir Potekhin",
```
Comment (splincode) [inline projects/core/package.json:22]: core?

Comment (vladimirpotekhin) [reply projects/core/package.json]: oooooops

File: projects/core/src/index.ts [hunk 15]
```diff
@@ -4,6 +4,7 @@ export {
 } from './lib/constants';
 export {Maskito} from './lib/mask';
 export {
+    MaskExpression,
```
Comment (nsbarsukov) [inline projects/core/src/index.ts:7]: Public API?

Let's rename it `MaskExpression` => `MaskitoMaskExpression`

File: projects/demo/src/pages/phone/examples/1-basic/component.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+            />
+        </tui-input>
+        <tui-error
+            [formControl]="value"
+            [error]="[] | tuiFieldError | async"
+        ></tui-error>
+    `,
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class DateMaskDocExample1 {
```
Comment (nsbarsukov) [inline projects/demo/src/pages/phone/examples/1-basic/component.ts:42]:
```suggestion
export class PhoneMaskDocExample1 {
```

File: projects/phone/package.json [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        "url": "https://github.com/vladimirpotekhin"
+    },
+    "contributors": [
+        "Alex Inkin <alexander@inkin.ru> (https://twitter.com/waterplea)",
+        "Maksim Ivanov <splincodewd@yandex.ru>",
+        "Vladimir Potekhin <vladimir.potekh@gmail.com>",
+        "Nikita Barsukov <nikita.s.barsukov@gmail.com>"
+    ],
+    "devDependencies": {
+        "libphonenumber-js": "^1.10.0"
```
Comment (nsbarsukov) [inline projects/phone/package.json:32]: Let's fixate this `devDependency` (**not `peerDependencies`!**) for renovate bot ?

File: projects/phone/project.json [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        },
+        "test": {
+            "executor": "@nrwl/jest:jest",
+            "outputs": ["coverage/phone"],
+            "options": {
+                "jestConfig": "projects/phone/jest.config.js",
+                "passWithNoTests": true
+            }
+        },
+        "publish": {
```
Comment (nsbarsukov) [inline projects/phone/project.json:39]: Should we drop this command for a while ?

File: projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts [hunk 19]
```diff
@@ -0,0 +1,20 @@
+import {MaskitoPostprocessor} from '@maskito/core';
+import {CountryCode, MetadataJson} from 'libphonenumber-js/core';
+
+import {cutPhoneByValidLength} from '../utils';
+
+const MIN_LENGTH = 3;
+export function maskitoCutPhonePostprocessorGenerator(
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts:7]: What do you think about this naming ?

```suggestion
export function maskitoPhoneLengthPostprocessorGenerator(
```

File: projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+): MaskitoPostprocessor {
+    return ({value, selection}) => {
+        return {
+            value:
+                value.length > MIN_LENGTH
+                    ? cutPhoneByValidLength({phone: value, countryIsoCode, metadata})
+                    : value,
+            selection,
+        };
+    };
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts:19]:
```suggestion
    return ({value, selection}) => ({
            value:
                value.length > MIN_LENGTH
                    ? cutPhoneByValidLength({phone: value, countryIsoCode, metadata})
                    : value,
            selection,
        });
```

File: projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts (hunk 19 above)
Comment (nsbarsukov) [reply projects/phone/src/lib/masks/phone/processors/cut-phone-postprocessor.ts:7]: \+ new file name `phone-length-postprocessor.ts`

File: projects/phone/src/lib/masks/phone/utils/generate-phone-mask.ts [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    value,
+    template,
+    prefix,
+}: {
+    value: string;
+    template: string;
+    prefix: string;
+}): MaskExpression {
+    return [
+        ...prefix.split(''),
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/utils/generate-phone-mask.ts:15]:
```suggestion
        ...prefix,
```

```
[...'abc'] // ['a', 'b', 'c']
```

File: projects/phone/src/lib/masks/phone/utils/get-phone-template.ts [hunk 22]
```diff
@@ -0,0 +1,16 @@
+import {AsYouType} from 'libphonenumber-js/core';
+
+export function getPhoneTemplate(formatter: AsYouType, value: string): string {
+    formatter.input(value.replace(/[\s\-()[\]{}]/g, ''));
```
Comment (nsbarsukov) [inline projects/phone/src/lib/masks/phone/utils/get-phone-template.ts:4]: Could it be replaces by the following line ?

```suggestion
    formatter.input(value.replace(/\D/g, ''));
```

File: projects/phone/src/lib/plugins/add-on-focus.ts [hunk 23]
```diff
@@ -0,0 +1,11 @@
+import {MaskitoPlugin} from '@maskito/core';
+import {maskitoEventHandler} from '@maskito/kit';
+
+export function maskitoAddOnFocusPlugin(value: string): MaskitoPlugin {
```
Comment (nsbarsukov) [inline projects/phone/src/lib/plugins/add-on-focus.ts:4]: Should we move `maskitoAddOnFocusPlugin` / `maskitoRemoveOnBlurPlugin` to `@maskito/kit` ? 🤔

File: projects/demo-integrations/cypress/tests/addons/phone/phone-basic.cy.ts [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+                        cy.get('@input')
+                            .type('5')
+                            .should('have.value', '+7 920 424-53-2')
+                            .should('have.prop', 'selectionStart', '+7 920 424-5'.length)
+                            .should('have.prop', 'selectionEnd', '+7 920 424-5'.length);
+                    });
+
+                    it('+7 920 424-11-32 => Select "+7 920 424-1|1-3|2" => Type "5" => +7 920 424-15-|2', () => {
+                        cy.get('@input')
+                            .type('{leftArrow}'.repeat('2'.length))
```
Comment (nsbarsukov) [inline projects/demo-integrations/cypress/tests/addons/phone/phone-basic.cy.ts:153]:
```suggestion
                            .type('{leftArrow}')
```

File: projects/phone/src/lib/plugins/add-on-focus.ts (hunk 23 above)
Comment (vladimirpotekhin) [reply projects/phone/src/lib/plugins/add-on-focus.ts]: I wanted to suggest the same

File: projects/phone/src/lib/masks/phone/utils/get-phone-template.ts (hunk 22 above)
Comment (vladimirpotekhin) [reply projects/phone/src/lib/masks/phone/utils/get-phone-template.ts:4]: no, we need "+" char

File: projects/phone/src/lib/masks/phone/utils/generate-phone-mask.ts (hunk 21 above)
Comment (vladimirpotekhin) [reply projects/phone/src/lib/masks/phone/utils/generate-phone-mask.ts]: It requires  `"downlevelIteration": true` in ts-config. I added

File: projects/demo/src/pages/phone/phone-doc.component.ts [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+type GeneratorOptions = Required<Parameters<typeof maskitoPhoneOptionsGenerator>[0]>;
+
+@Component({
+    selector: 'phone-doc',
+    templateUrl: './phone-doc.template.html',
+    styleUrls: ['./phone-doc.style.less'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class PhoneDocComponent implements GeneratorOptions {
```
Comment (splincode) [inline projects/demo/src/pages/phone/phone-doc.component.ts:17]: what about example

```ts
async ngOnInit() {
    this.mask = maskitoPhoneOptionsGenerator({
      countryIsoCode: 'KZ',
      metadata: await import('libphonenumber-js/min/metadata')
          .then(({ default }) => default),
    });
```

File: projects/phone/src/lib/masks/phone/utils/get-phone-template.ts (hunk 22 above)
Comment (nsbarsukov) [reply projects/phone/src/lib/masks/phone/utils/get-phone-template.ts:4]: what about this option ?

```suggestion
    formatter.input(value.replace(/[^\d\+]/g, ''));
```

File: projects/demo/src/pages/phone/examples/1-basic/component.ts [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+                tuiTextfield
+                autocomplete="tel"
+                inputmode="tel"
+                [maskito]="mask"
+            />
+        </tui-input>
+        <tui-error
+            [formControl]="value"
+            [error]="[] | tuiFieldError | async"
+        ></tui-error>
```
Comment (nsbarsukov) [inline projects/demo/src/pages/phone/examples/1-basic/component.ts:25]: Do we need it for this example ?

File: projects/demo/src/pages/phone/examples/1-basic/component.ts [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+    `,
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class PhoneMaskDocExample1 {
+    value = new FormControl('+7 771 931-1111');
```
Comment (nsbarsukov) [inline projects/demo/src/pages/phone/examples/1-basic/component.ts:30]: `control ` ? Or `[(ngModel)]`?

File: projects/demo/src/pages/phone/examples/2-validation/component.ts [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+        </tui-input>
+        <tui-error
+            [formControl]="value"
+            [error]="[] | tuiFieldError | async"
+        ></tui-error>
+    `,
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class PhoneMaskDocExample2 {
+    value = new FormControl('+36 20 123-3122', phoneValidator('HU'));
```
Comment (nsbarsukov) [inline projects/demo/src/pages/phone/examples/2-validation/component.ts:45]:
```suggestion
    readonly control = new FormControl('+36 20 123-3122', phoneValidator('HU'));
```

File: projects/demo/src/pages/phone/phone-doc.template.html [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+        <tui-doc-example
+            id="validation"
+            heading="validation"
+            [description]="description"
+            [content]="validation"
+        >
+            <phone-doc-example-2></phone-doc-example-2>
+
+            <ng-template #description>
+                <p>
```
Comment (nsbarsukov) [inline projects/demo/src/pages/phone/phone-doc.template.html:38]:
```suggestion
                <p class="tui-space_top-0">
```

File: projects/demo/src/pages/phone/phone-doc.template.html [hunk 30]
```diff
@@ -0,0 +1,98 @@
...
+        <tui-doc-documentation>
+            <ng-template
+                documentationPropertyName="countryCode"
+                documentationPropertyMode="input"
+                documentationPropertyType="string"
+                [documentationPropertyValues]="countryCodeVariants"
+                [(documentationPropertyValue)]="countryIsoCode"
+                (documentationPropertyValueChange)="updateOptions()"
+            >
+                Country code
```
Comment (nsbarsukov) [inline projects/demo/src/pages/phone/phone-doc.template.html:94]:
```suggestion
                Country ISO-code
```
?

File: projects/phone/src/index.ts [hunk 31]
```diff
@@ -0,0 +1 @@
+export * from './lib/masks';
```
Comment (nsbarsukov) [inline projects/phone/src/index.ts:1]: Should we narrow public API ?

For example, i think that `TEMPLATE_FILLER` should not be in public API

Comment (nsbarsukov) [reply projects/phone/src/index.ts:1]: Should processors be in public API?

File: projects/demo/src/pages/phone/phone-doc.component.ts (hunk 25 above)
Comment (vladimirpotekhin) [reply projects/demo/src/pages/phone/phone-doc.component.ts:17]: An example with lazy metadata will be added to a separate PR

**Review by nsbarsukov — APPROVED** [review]

**Pull request merged**